engine/crates/composition/src/ingest_subgraph/directives.rs

```diff
@@ -109,3 +125,358 @@ pub(super) fn ingest_keys(
         }
     }
 }
+
+pub(super) fn ingest_directive_definitions<'a>(
+    document: &'a ast::ServiceDocument,
+    mut push_error: impl FnMut(String),
+) -> DirectiveMatcher<'a> {
+    let schema_definition_directives = document
+        .definitions
+        .iter()
+        .filter_map(|def| {
+            if let ast::TypeSystemDefinition::Schema(schema) = def {
+                Some(schema)
+            } else {
+                None
+            }
+        })
+        .flat_map(|definition| definition.node.directives.iter());
+
+    let mut directive_matcher = schema_definition_directives
+        .clone()
+        .map(|d| &d.node)
+        .find(|d| DirectiveMatcher::is_federation_directive(d))
+        .map(DirectiveMatcher::new)
+        .unwrap_or_default();
+
+    let mut composed_directives = BTreeSet::new();
+
+    for name in schema_definition_directives
+        .filter(|directive| directive_matcher.is_compose_directive(directive.node.name.node.as_str()))
+        .filter_map(|directive| directive.node.get_argument("name"))
+        .filter_map(|directive_name| match &directive_name.node {
+            ConstValue::String(s) => Some(s.as_str()),
+            _ => None,
+        })
+    {
+        composed_directives.insert(name.trim_start_matches('@'));
+
+        if !name.starts_with('@') {
+            push_error(format!(
+                "The `{}` directive is missing the `@` prefix in @composeDirective.",
+                name
+            ));
+        }
+    }
+
+    directive_matcher.composed_directives = composed_directives;
+
+    directive_matcher
+}
+
+/// This struct is the source of truth for matching federation directives by name when ingesting a
```

Comment:
From here, this file is 99% moved. Changes: removed FederationDirectivesMatcher to DirectiveMatcher and added the two struct fields about composed directives.